salt/engines/libvirt_events.py

```diff
@@ -0,0 +1,617 @@
+
+CALLBACK_DEFS = {
+    'domain': (('lifecycle', None),
+               ('reboot', None),
+               ('rtc_change', None),
+               ('watchdog', None),
+               ('graphics', None),
+               ('io_error', 'VIR_DOMAIN_EVENT_ID_IO_ERROR_REASON'),
+               ('control_error', None),
+               ('disk_change', None),
+               ('tray_change', None),
+               ('pmwakeup', None),
+               ('pmsuspend', None),
+               ('balloon_change', None),
+               ('pmsuspend_disk', None),
+               ('device_removed', None),
+               ('block_job', BLOCK_JOB_ID),
+               ('tunable', None),
+               ('agent_lifecycle', None),
+               ('device_added', None),
+               ('migration_iteration', None),
+               ('job_completed', None),
+               ('device_removal_failed', None),
+               ('metadata_change', None),
+               ('block_threshold', None)),
+    'network': (('lifecycle', None),),
+    'pool': (('lifecycle', None),
+             ('refresh', None)),
+    'nodedev': (('lifecycle', None),
+                ('update', None)),
+    'secret': (('lifecycle', None),
+               ('value_changed', None))
+}
+
+
+def _compute_subprefix(attr):
+    '''
+    Get the part before the first '_' or the end of @attr including
+    the potential '_'
+    '''
+    return ''.join((attr.split('_')[0], '_' if len(attr.split('_')) > 1 else ''))
+
+
+def _get_libvirt_enum_string(prefix, value):
+    '''
+    Convert the libvirt enum integer value @value into a human readable string.
+    @prefix is the start of the libvirt attribute to look for.
+    '''
+    attributes = [attr[len(prefix):] for attr in libvirt.__dict__ if attr.startswith(prefix)]
+
+    # Filter out the values starting with a common base as they match another enum
+    prefixes = [_compute_subprefix(p) for p in attributes]
+    counts = {p: prefixes.count(p) for p in prefixes}
+    sub_prefixes = [p for p, count in counts.items() if count > 1]
+    filtered = [attr for attr in attributes if _compute_subprefix(attr) not in sub_prefixes]
+
+    for candidate in filtered:
+        if value == getattr(libvirt, ''.join((prefix, candidate))):
+            name = candidate.lower().replace('_', ' ')
+            return name
+    return 'unknown'
+
+
+def _get_domain_event_detail(event, detail):
+    '''
+    Convert @event and @detail numeric values into a tuple of human readable strings
+    '''
+    event_name = _get_libvirt_enum_string('VIR_DOMAIN_EVENT_', event)
+    if event_name == 'unknown':
+        return event_name, 'unknown'
+
+    prefix = 'VIR_DOMAIN_EVENT_{0}_'.format(event_name.upper())
+    detail_name = _get_libvirt_enum_string(prefix, detail)
+
+    return event_name, detail_name
+
+
+def salt_send_event(opaque, conn, data):
+    '''
+    Convenience function adding common data to the event and sending it
+    on the salt event bus.
+    '''
```

Comment:
Can you document all of the arguments here? And provide an example of how they should be used?